File: examples/java/src/main/java/io/superdurable/dex/patterns/parallelsubflows/AdvancedLongLiveParentFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 */

package io.superdurable.dex.patterns.parallelsubflows;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.SubFlow;
import io.superdurable.dex.Wait;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public final class AdvancedLongLiveParentFlow implements Flow<ParentInput> {
    public static final int DEFAULT_CONCURRENCY = 10;
    public static final int MAX_BUFFERED_REQUESTS = 100;

    public final Channel<String> requestChannel = Channel.define("RequestChannel", String.class);
    public final Attribute<Boolean> stopped = Attribute.define("Stopped", Boolean.class);

    private final InitStep initStep = new InitStep();
    private final HandleRequestStep handleRequestStep = new HandleRequestStep();
    private final HandleSubFlowStep handleSubFlowStep;

    public AdvancedLongLiveParentFlow(final ExampleSubFlow exampleSubFlow) {
        handleSubFlowStep = new HandleSubFlowStep(exampleSubFlow);
    }

    @Override
    public StepList<ParentInput> getSteps() {
        return StepList.startStep(initStep).otherSteps(handleRequestStep, handleSubFlowStep);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(requestChannel, stopped);
    }

    @RPC
    public RPCResult<Boolean> sendRequest(final Context context, final String request) {
        if (requestChannel.size(context) >= MAX_BUFFERED_REQUESTS) {
            return RPCResult.of(false);
        }
        requestChannel.publish(context, request);
        return RPCResult.of(true);
    }

    @RPC
    public void stop(final Context context) {
        stopped.set(context, true);
    }

    final class InitStep implements Step<ParentInput> {
        @Override
        public Class<ParentInput> getInputType() {
            return ParentInput.class;
        }

        @Override
        public StepDecision execute(final Context context, final ParentInput input) {
            for (final String request : input.requests) {
                requestChannel.publish(context, request);
            }
            stopped.set(context, false);
            final int concurrency = input.concurrency > 0 ? input.concurrency : DEFAULT_CONCURRENCY;
            final StepMovement<?>[] movements = new StepMovement<?>[concurrency];
            for (int index = 0; index < concurrency; index++) {
                movements[index] = StepMovement.of(HandleRequestStep.class, null);
            }
            return StepDecision.goToMany(movements);
        }
    }

    final class HandleRequestStep implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public Wait waitFor(final Context context, final Void input) {
            return Wait.until(requestChannel.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final List<String> requests = requestChannel.getConditionResults(context);
            return StepDecision.goTo(HandleSubFlowStep.class, requests.get(0));
        }
    }

    final class HandleSubFlowStep implements Step<String> {
        private final ExampleSubFlow exampleSubFlow;

        HandleSubFlowStep(final ExampleSubFlow exampleSubFlow) {
            this.exampleSubFlow = exampleSubFlow;
        }

        @Override
        public Class<String> getInputType() {
            return String.class;
        }

        @Override
        public Wait waitFor(final Context context, final String request) {
            return Wait.until(SubFlow.run(ExampleSubFlow.class, request));
        }

        @Override
        public StepDecision execute(final Context context, final String request) {
            if (Boolean.TRUE.equals(stopped.get(context))) {
                return StepDecision.gracefulComplete(null);
            }
            return StepDecision.goTo(HandleRequestStep.class, null);
        }
    }
}

File: examples/go/patterns/parallel-subflows/advanced_long_live_parent_flow.go
// Copyright (c) 2022-2026 Super Durable, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package parallelsubflows

import "github.com/superdurable/dex/sdk-go/dex"

var (
	LongLiveRequestChannel = dex.DefineChannel[string]("RequestChannel")
	Stopped                = dex.DefineAttribute[bool]("Stopped")
)

type AdvancedLongLiveParentFlow struct {
	dex.FlowDefaults
	exampleFlow *ExampleSubFlow
}

func NewAdvancedLongLiveParentFlow(exampleFlow *ExampleSubFlow) *AdvancedLongLiveParentFlow {
	return &AdvancedLongLiveParentFlow{exampleFlow: exampleFlow}
}

func (flow *AdvancedLongLiveParentFlow) GetSteps() []dex.StepDef {
	return []dex.StepDef{
		dex.DefineStartStep(longLiveInitStep{}),
		dex.DefineStep(longLiveHandleRequestStep{}),
		dex.DefineStep(longLiveHandleSubFlowStep{exampleFlow: flow.exampleFlow}),
	}
}

func (*AdvancedLongLiveParentFlow) GetPersistenceSchema() dex.PersistenceSchema {
	return dex.PersistenceSchema{
		Attributes: []dex.AttributeDef{Stopped},
		Channels:   []dex.ChannelDef{LongLiveRequestChannel},
	}
}

func (*AdvancedLongLiveParentFlow) SendRequest(ctx dex.Context, request string) (*dex.RPCResult[bool], error) {
	if LongLiveRequestChannel.Size(ctx) >= MaxBufferedRequests {
		return &dex.RPCResult[bool]{Output: false}, nil
	}
	if err := LongLiveRequestChannel.Publish(ctx, request); err != nil {
		return nil, err
	}
	return &dex.RPCResult[bool]{Output: true}, nil
}

func (*AdvancedLongLiveParentFlow) Stop(ctx dex.Context, _ dex.None) (*dex.RPCResult[dex.None], error) {
	if err := Stopped.Set(ctx, true); err != nil {
		return nil, err
	}
	return &dex.RPCResult[dex.None]{}, nil
}

type longLiveInitStep struct {
	dex.StepDefaultsNoWaitFor[ParentInput]
}

func (longLiveInitStep) GetStepType() string { return "InitStep" }

func (longLiveInitStep) Execute(ctx dex.Context, input ParentInput) (*dex.StepDecision, error) {
	for _, request := range input.Requests {
		if err := LongLiveRequestChannel.Publish(ctx, request); err != nil {
			return nil, err
		}
	}
	if err := Stopped.Set(ctx, false); err != nil {
		return nil, err
	}
	count := concurrency(input.Concurrency)
	movements := make([]dex.StepMovement, 0, count)
	for index := 0; index < count; index++ {
		movements = append(movements, dex.MovementOf(longLiveHandleRequestStep{}, nil))
	}
	return dex.GoToMany(movements...), nil
}

type longLiveHandleRequestStep struct{ dex.StepDefaults }

func (longLiveHandleRequestStep) GetStepType() string { return "HandleRequestStep" }

func (longLiveHandleRequestStep) WaitFor(_ dex.Context, _ dex.None) (*dex.Wait, error) {
	return dex.Until(LongLiveRequestChannel.ForOne()), nil
}

func (longLiveHandleRequestStep) Execute(ctx dex.Context, _ dex.None) (*dex.StepDecision, error) {
	requests, err := LongLiveRequestChannel.GetConditionResults(ctx)
	if err != nil {
		return nil, err
	}
	return dex.GoTo(longLiveHandleSubFlowStep{}, requests[0]), nil
}

type longLiveHandleSubFlowStep struct {
	dex.StepDefaults
	exampleFlow *ExampleSubFlow
}

func (longLiveHandleSubFlowStep) GetStepType() string { return "HandleSubFlowStep" }

func (step longLiveHandleSubFlowStep) WaitFor(_ dex.Context, request string) (*dex.Wait, error) {
	return dex.Until(dex.SubFlow(step.exampleFlow, request)), nil
}

func (longLiveHandleSubFlowStep) Execute(ctx dex.Context, _ string) (*dex.StepDecision, error) {
	stopped, err := Stopped.Get(ctx)
	if err != nil {
		return nil, err
	}
	if stopped {
		return dex.GracefulComplete(nil), nil
	}
	return dex.GoTo(longLiveHandleRequestStep{}, nil), nil
}

var (
	_ dex.Flow                    = (*AdvancedLongLiveParentFlow)(nil)
	_ dex.RPC[string, bool]       = (*AdvancedLongLiveParentFlow)(nil).SendRequest
	_ dex.RPC[dex.None, dex.None] = (*AdvancedLongLiveParentFlow)(nil).Stop
)
